src/core/useAtom.ts

```diff
@@ -38,85 +22,8 @@ export function useAtom<Value, Update, Result extends void | Promise<void>>(
     )
     scope = (atom as { scope: Scope }).scope
   }
-
-  const ScopeContext = getScopeContext(scope)
-  const { s: store, w: versionedWrite } = useContext(ScopeContext)
-
-  const getAtomValue = useCallback(
-    (version?: VersionObject) => {
-      // This call to READ_ATOM is the place where derived atoms will actually be
-      // recomputed if needed.
-      const atomState = store[READ_ATOM](atom, version)
-      if ('e' in atomState) {
-        throw atomState.e // read error
-      }
-      if ('p' in atomState) {
-        throw atomState.p // read promise
-      }
-      if ('v' in atomState) {
-        return atomState.v as ResolveType<Value>
-      }
-      throw new Error('no atom value')
-    },
-    [store, atom]
-  )
-
-  // Pull the atoms's state from the store into React state.
-  const [[version, value, atomFromUseReducer], rerenderIfChanged] = useReducer<
-    Reducer<
-      readonly [VersionObject | undefined, ResolveType<Value>, Atom<Value>],
-      VersionObject | undefined
-    >,
-    undefined
-  >(
-    useCallback(
-      (prev, nextVersion) => {
-        const nextValue = getAtomValue(nextVersion)
-        if (Object.is(prev[1], nextValue) && prev[2] === atom) {
-          return prev // bail out
-        }
-        return [nextVersion, nextValue, atom]
-      },
-      [getAtomValue, atom]
-    ),
-    undefined,
-    () => {
-      // NOTE should/could branch on mount?
-      const initialVersion = undefined
-      const initialValue = getAtomValue(initialVersion)
-      return [initialVersion, initialValue, atom]
-    }
-  )
-
-  if (atomFromUseReducer !== atom) {
-    rerenderIfChanged(undefined)
-  }
-
-  useEffect(() => {
-    // Call `rerenderIfChanged` whenever this atom is invalidated. Note
-    // that derived atoms may not be recomputed yet.
-    const unsubscribe = store[SUBSCRIBE_ATOM](atom, rerenderIfChanged)
-    rerenderIfChanged(undefined)
-    return unsubscribe
-  }, [store, atom])
-
-  useEffect(() => {
-    store[COMMIT_ATOM](atom, version)
-  })
-
-  const setAtom = useCallback(
-    (update: Update) => {
-      if (isWritable(atom)) {
-        const write = (version?: VersionObject) =>
-          store[WRITE_ATOM](atom, update, version)
-        return versionedWrite ? versionedWrite(write) : write()
-      } else {
-        throw new Error('not writable atom')
-      }
-    },
-    [store, versionedWrite, atom]
-  )
-
-  useDebugValue(value)
-  return [value, setAtom]
+  return [
+    useAtomValue(atom, scope),
+    useSetAtom(atom as WritableAtom<Value, Update, Result>, scope),
+  ]
```

Comment:
This is what I expected, but now after reviewing this commit, I think checking `isWritable` here seems better.
```suggestion
  return [
    useAtomValue(atom, scope),
    isWritable(atom) ? useSetAtom(atom, scope) : () => {
      throw new Error('not writable atom')
    },
  ]
```

Reply:
Oh my bad, this breaks the rules of hooks. Void.